{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load libraries\n",
    "from IPython.display import display, HTML\n",
    "import pandas as pd\n",
    "import polars as pl\n",
    "import scipy\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.patches import Rectangle\n",
    "from datetime import date\n",
    "from pathlib import Path\n",
    "import gzip\n",
    "import re\n",
    "\n",
    "import os\n",
    "import subprocess\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "version = %env WORKSPACE_CDR\n",
    "my_bucket = os.getenv('WORKSPACE_BUCKET')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Polars string length to 100\n",
    "pl.Config.set_fmt_str_lengths(100)\n",
    "\n",
    "# Set the row limit to a higher value\n",
    "pl.Config.set_tbl_rows(50)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gwaslab as gl"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gene Annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hla_start = 29602238\n",
    "hla_end = 33409896"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_gene_annotations(gtf_path, chromosome, start_pos, end_pos, \n",
    "                           build=\"hg38\", protein_coding_only=False):\n",
    "    \"\"\"Extract gene annotations for Manhattan plot\"\"\"\n",
    "    from gwaslab.viz_plot_regionalplot import process_gtf\n",
    "    \n",
    "    # Set up parameters like gwaslab does\n",
    "    region = [chromosome, start_pos, end_pos]\n",
    "    gtf_chr_dict = {chromosome: str(chromosome)}  # or \"chr\" + str(chromosome)\n",
    "    \n",
    "    uniq_genes, exons = process_gtf(\n",
    "        gtf_path=gtf_path,\n",
    "        region=region,\n",
    "        region_flank_factor=0.1,  # 10% flank\n",
    "        build=build,\n",
    "        region_protein_coding=protein_coding_only,\n",
    "        gtf_chr_dict=gtf_chr_dict,\n",
    "        gtf_gene_name='gene_name'\n",
    "    )\n",
    "    \n",
    "    # Return simplified gene annotation DataFrame\n",
    "    gene_annotations = uniq_genes[['name', 'start', 'end']].copy()\n",
    "    gene_annotations['mid_point'] = (gene_annotations['start'] + gene_annotations['end']) // 2\n",
    "    \n",
    "    return gene_annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hla_region_genes = pl.from_pandas(extract_gene_annotations(\n",
    "    gtf_path=gl.get_path('ensembl_hg38_gtf'),\n",
    "    chromosome=6,\n",
    "    start_pos=hla_start,  # HLA region start\n",
    "    end_pos=hla_end,    # HLA region end\n",
    "    build=\"hg38\",\n",
    "    protein_coding_only=True\n",
    "))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hla_region_genes = pl.concat([\n",
    "    hla_region_genes,\n",
    "    pl.DataFrame({\n",
    "        'name': ['HLA-H', 'HLA-J', 'HLA-L', 'HLA-DRB3'],\n",
    "        'start': [29887752, 30006606, 30259625, 32345227], \n",
    "        'end': [29890482, 30009539, 30261703, 32358367],\n",
    "        'mid_point': [29889117, 30008072, 30260664, 32351797]\n",
    "    })\n",
    "]).sort('start')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hla_region_genes.filter(pl.col('name').str.contains('HLA'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Anal Disease"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## HLA and GWAS Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "afr_anal_hla_df = pl.read_csv(f'{my_bucket}/HLA/HPV_AFR_HLA_results_freq.csv')\n",
    "eur_anal_hla_df = pl.read_csv(f'{my_bucket}/HLA/HPV_EUR_HLA_results_freq.csv')\n",
    "amr_anal_hla_df = pl.read_csv(f'{my_bucket}/HLA/HPV_AMR_HLA_results_freq.csv')\n",
    "metal_anal_hla_df = pl.read_csv(f'{my_bucket}/HLA/meta_HPV_1.txt', separator='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "eur_anal_sumstats = gl.Sumstats(f'{my_bucket}/saige_gwas/v1_redo/eur/condition__hpv/gwas_results.tsv.gz',\n",
    "                           fmt=\"saige\",\n",
    "                           snpid='vid',\n",
    "                           rsid=None,\n",
    "                           chrom='chromosome',\n",
    "                           pos='base_pair_location',\n",
    "                           ea='effect_allele',\n",
    "                           nea='other_allele',\n",
    "                           eaf='effect_allele_frequency',\n",
    "                           se='standard_error',\n",
    "                           p='p_value',\n",
    "                           beta='beta',\n",
    "                           build='38'\n",
    ")\n",
    "eur_anal_sumstats.basic_check()\n",
    "eur_hla_anal_sumstats = pl.from_pandas(eur_anal_sumstats.data).filter(\n",
    "    (pl.col('CHR')==6) &\n",
    "    ((pl.col('POS')>hla_start) | (pl.col('CHR')<hla_end))\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "afr_anal_sumstats = gl.Sumstats(f'{my_bucket}/saige_gwas/v1_redo/afr/condition__hpv/gwas_results.tsv.gz',\n",
    "                           fmt=\"saige\",\n",
    "                           snpid='vid',\n",
    "                           rsid=None,\n",
    "                           chrom='chromosome',\n",
    "                           pos='base_pair_location',\n",
    "                           ea='effect_allele',\n",
    "                           nea='other_allele',\n",
    "                           eaf='effect_allele_frequency',\n",
    "                           se='standard_error',\n",
    "                           p='p_value',\n",
    "                           beta='beta',\n",
    "                           build='38'\n",
    ")\n",
    "afr_anal_sumstats.basic_check()\n",
    "afr_hla_anal_sumstats = pl.from_pandas(afr_anal_sumstats.data).filter(\n",
    "    (pl.col('CHR')==6) &\n",
    "    ((pl.col('POS')>hla_start) | (pl.col('CHR')<hla_end))\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "amr_anal_sumstats = gl.Sumstats(f'{my_bucket}/saige_gwas/v1_redo/amr/condition__hpv/gwas_results.tsv.gz',\n",
    "                           fmt=\"saige\",\n",
    "                           snpid='vid',\n",
    "                           rsid=None,\n",
    "                           chrom='chromosome',\n",
    "                           pos='base_pair_location',\n",
    "                           ea='effect_allele',\n",
    "                           nea='other_allele',\n",
    "                           eaf='effect_allele_frequency',\n",
    "                           se='standard_error',\n",
    "                           p='p_value',\n",
    "                           beta='beta',\n",
    "                           build='38'\n",
    ")\n",
    "amr_anal_sumstats.basic_check()\n",
    "amr_hla_anal_sumstats = pl.from_pandas(amr_anal_sumstats.data).filter(\n",
    "    (pl.col('CHR')==6) &\n",
    "    ((pl.col('POS')>hla_start) | (pl.col('CHR')<hla_end))\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_chr_pos_from_snpid(df: pd.DataFrame, snp_col: str = \"SNPID\") -> pd.DataFrame:\n",
    "    \"\"\"Extract CHR and POS from SNPID column (format CHR-POS-...).\"\"\"\n",
    "    parts = df[snp_col].str.split(\"-\", n=2, expand=True)\n",
    "    df = df.copy()\n",
    "    df[\"CHR\"] = pd.to_numeric(parts[0], errors=\"coerce\")\n",
    "    df[\"POS\"] = pd.to_numeric(parts[1], errors=\"coerce\")\n",
    "    return df\n",
    "\n",
    "metal_anal_sumstats = gl.Sumstats(f'{my_bucket}/saige_gwas/v1_redo/metal/condition__hpv/condition__hpv_afr_amr_eur1.tsv',\n",
    "                           fmt=\"metal\",\n",
    "                           snpid='MarkerName',\n",
    "                           rsid=None,\n",
    "                           ea='Allele1',\n",
    "                           nea='Allele2',\n",
    "                           eaf='Freq1',\n",
    "                           se='StdErr',\n",
    "                           p='P-value',\n",
    "                           beta='Effect',\n",
    "                           direction='Direction',\n",
    "                           build='38'\n",
    ")\n",
    "# Get CHR POS from SNPID\n",
    "metal_anal_sumstats.data = add_chr_pos_from_snpid(metal_anal_sumstats.data, snp_col='SNPID')\n",
    "metal_anal_sumstats.basic_check()\n",
    "metal_hla_anal_sumstats = pl.from_pandas(metal_anal_sumstats.data).filter(\n",
    "    (pl.col('CHR')==6) &\n",
    "    ((pl.col('POS')>hla_start) | (pl.col('CHR')<hla_end))\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metal_anal_hla_df = metal_anal_hla_df.rename({'MarkerName':'Marker', 'P-value': 'P'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hla_genes_positions = hla_region_genes.filter(pl.col('name').str.contains('HLA')).sort('name')\n",
    "    \n",
    "hla_genes_positions = hla_genes_positions.with_columns(\n",
    "    pl.col(\"name\").str.split(\"-\").list.last().alias(\"hla_name\"),\n",
    ")\n",
    "\n",
    "locus_positions = dict(zip(hla_genes_positions['hla_name'], hla_genes_positions['mid_point']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove DMA and DRA from analysis because AC == 1 for both\n",
    "eur_anal_hla_df = eur_anal_hla_df.filter(~pl.col('Marker').str.contains('DMA|DRA'))\n",
    "afr_anal_hla_df = afr_anal_hla_df.filter(~pl.col('Marker').str.contains('DMA|DRA'))\n",
    "amr_anal_hla_df = amr_anal_hla_df.filter(~pl.col('Marker').str.contains('DMA|DRA'))\n",
    "metal_anal_hla_df = metal_anal_hla_df.filter(~pl.col('Marker').str.contains('DMA|DRA'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Prepare your data dictionary\n",
    "data_dict = {\n",
    "    'eur': (eur_anal_hla_df, eur_hla_anal_sumstats),\n",
    "    'afr': (afr_anal_hla_df, afr_hla_anal_sumstats), \n",
    "    'amr': (amr_anal_hla_df, amr_hla_anal_sumstats),\n",
    "    'metal': (metal_anal_hla_df, metal_hla_anal_sumstats)\n",
    "}\n",
    "\n",
    "# Calculate Bonferroni thresholds\n",
    "bonf_thresholds = {\n",
    "    'eur': 0.05 / len(eur_anal_hla_df),\n",
    "    'afr': 0.05 / len(afr_anal_hla_df),\n",
    "    'amr': 0.05 / len(amr_anal_hla_df),\n",
    "    'metal': 0.05 / len(metal_anal_hla_df)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### GWAS Overlay"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.colors as mcolors\n",
    "\n",
    "def create_hla_manhattan_plot(data_dict, hla_region_genes, locus_positions, bonf_thresholds, figsize=(12, 12)):\n",
    "    \"\"\"\n",
    "    Create a multi-panel HLA Manhattan plot with gene annotations\n",
    "    \"\"\"\n",
    "    \n",
    "    # Define HLA class I and II genes for color mapping\n",
    "    hla_class_i = ['A', 'B', 'C', 'E', 'F', 'G', 'H', 'J', 'L']\n",
    "    hla_class_ii = [ 'DMB', 'DOA', 'DOB', 'DPA1',                         # 'DMA',\n",
    "                    'DPB1', 'DQA1', 'DQB1', 'DRB1', 'DRB3', 'DRB5']       #'DRA', \n",
    " \n",
    "    # Color palettes for HLA classes - using darker colors\n",
    "    class_i_colors = sns.color_palette(\"Reds\", n_colors=len(hla_class_i)+3)[3:]  # Skip 3 lightest\n",
    "    class_ii_colors = sns.color_palette(\"Blues\", n_colors=len(hla_class_ii)+3)[3:]  # Skip 3 lightest\n",
    "    \n",
    "    # Gene → color mapping\n",
    "    hla_gene_colors = {g: c for g, c in zip(hla_class_i, class_i_colors)}\n",
    "    hla_gene_colors.update({g: c for g, c in zip(hla_class_ii, class_ii_colors)})\n",
    "    hla_gene_colors = {g: mcolors.to_hex(c) for g, c in hla_gene_colors.items()}\n",
    "\n",
    "    # Hard-coded label positions to avoid overlap\n",
    "    label_positions = {\n",
    "        'HLA-F': 29.728,      \n",
    "        'HLA-G': 29.825,      \n",
    "        'HLA-H': 29.887,      \n",
    "        'HLA-A': 29.95,      \n",
    "        'HLA-J': 30.005,      \n",
    "        'HLA-B': 31.36,        \n",
    "        'HLA-C': 31.27,      \n",
    "#         'HLA-E': 30.495,      \n",
    "        'HLA-L': 30.26,      \n",
    "        'HLA-DRB3': 32.325,   \n",
    "#         'HLA-DRA': 32.413,    \n",
    "        'HLA-DRB5': 32.498,   \n",
    "        'HLA-DRB1': 32.565,   \n",
    "        'HLA-DQA1': 32.638,   \n",
    "        'HLA-DQB1': 32.688,   \n",
    "        'HLA-DOB': 32.815,    \n",
    "        'HLA-DMB': 32.905,    \n",
    "#         'HLA-DMA': 32.962,    \n",
    "        'HLA-DOA': 33.008,    \n",
    "        'HLA-DPA1': 33.125,   \n",
    "        'HLA-DPB1': 33.118,   \n",
    "    }\n",
    "    \n",
    "    # Create figure with subplots\n",
    "    fig, axes = plt.subplots(4, 1, figsize=figsize, sharex=True)\n",
    "    fig.suptitle('HLA Region Association Results by Ancestry', fontsize=16, y=0.98)\n",
    "    \n",
    "    ancestries = ['eur', 'afr', 'amr', 'metal']\n",
    "    ancestry_labels = ['EUR', 'AFR', 'AMR', 'METAL']\n",
    "    \n",
    "    for idx, (ancestry, label) in enumerate(zip(ancestries, ancestry_labels)):\n",
    "        ax = axes[idx]\n",
    "        hla_results, hla_anal_sumstats = data_dict[ancestry]\n",
    "        bonferroni = bonf_thresholds[ancestry]\n",
    "        \n",
    "        # Plot background GWAS results\n",
    "        ax.scatter(hla_anal_sumstats['POS'], -np.log10(hla_anal_sumstats['P']),\n",
    "                   c='lightgray', alpha=0.4, s=8, zorder=1)\n",
    "\n",
    "        df = (\n",
    "            hla_results\n",
    "            .with_columns([\n",
    "                pl.col(\"Marker\").str.split(\".\").list.first().alias(\"HLA_gene\"),\n",
    "                pl.when(pl.col(\"P\") > 0)\n",
    "                  .then(-pl.col(\"P\").log10())\n",
    "                  .otherwise(None)\n",
    "                  .alias(\"neg_log10P\"),\n",
    "            ])\n",
    "            .with_columns(\n",
    "                pl.col(\"HLA_gene\").replace(locus_positions, default=None).alias(\"pos\")\n",
    "            )\n",
    "        )\n",
    "\n",
    "        # Map HLA gene → color/alpha/size\n",
    "        df = df.with_columns([\n",
    "            pl.col(\"HLA_gene\").replace(hla_gene_colors, default=\"darkgray\").alias(\"color\"),\n",
    "            pl.when(pl.col(\"HLA_gene\").is_in(list(hla_gene_colors)))\n",
    "              .then(1).otherwise(0.6).alias(\"alpha\"),\n",
    "            pl.when(pl.col(\"HLA_gene\").is_in(list(hla_gene_colors)))\n",
    "              .then(20).otherwise(8).alias(\"size\"),\n",
    "        ])\n",
    "\n",
    "        # Regular scatter plot \n",
    "        ax.scatter(\n",
    "            df[\"pos\"].to_numpy(),\n",
    "            df[\"neg_log10P\"].to_numpy(),\n",
    "            c=df[\"color\"].to_list(),\n",
    "            alpha=df[\"alpha\"].to_numpy(),\n",
    "            s=df[\"size\"].to_numpy(),\n",
    "            zorder=3,\n",
    "            edgecolors=\"white\",\n",
    "            linewidth=0.5\n",
    "        )\n",
    "\n",
    "        # Gene track below x-axis\n",
    "        gene_track_y = -2.5\n",
    "        \n",
    "        for gene in hla_region_genes.iter_rows(named=True):\n",
    "            gene_width = gene['end'] - gene['start']\n",
    "            if gene['name'].startswith('HLA-'):\n",
    "                hla_gene = gene['name']\n",
    "                if hla_gene in hla_gene_colors:\n",
    "                    color, alpha = hla_gene_colors[hla_gene], 0.8\n",
    "                else:\n",
    "                    color, alpha = 'gray', 0.6\n",
    "                    \n",
    "                rect = Rectangle((gene['start'], gene_track_y), gene_width, 0.6,\n",
    "                                 facecolor=color, alpha=alpha, zorder=2)\n",
    "                ax.add_patch(rect)\n",
    "                \n",
    "                # Draw label at fixed position if we have one\n",
    "                if hla_gene in label_positions:\n",
    "                    label_x = label_positions[hla_gene] * 1e6  # Convert Mb to bp\n",
    "                    original_x = gene['mid_point']\n",
    "                    \n",
    "                    # Check if it's an HLA-D gene for alternation logic\n",
    "                    if hla_gene.startswith('HLA-D'):\n",
    "                        # Alternate HLA-D genes: even indices on top, odd on bottom\n",
    "                        gene_idx = list(label_positions.keys()).index(hla_gene)\n",
    "                        if gene_idx % 2 == 0:  # Top\n",
    "                            label_y = gene_track_y + .9\n",
    "                            line_start_y = gene_track_y + 0.6\n",
    "                            line_end_y = gene_track_y + 0.8\n",
    "                            va = 'bottom'\n",
    "                        else:  # Bottom\n",
    "                            label_y = gene_track_y - .3\n",
    "                            line_start_y = gene_track_y\n",
    "                            line_end_y = gene_track_y - 0.4\n",
    "                            va = 'top'\n",
    "                        \n",
    "                        # Draw connecting line for HLA-D genes only\n",
    "                        ax.plot([original_x, label_x], \n",
    "                               [line_start_y, line_end_y], \n",
    "                               color='gray', linestyle='-', linewidth=0.5, alpha=0.7)\n",
    "                    else:\n",
    "                        # All non-HLA-D genes go to bottom (no connecting lines)\n",
    "                        label_y = gene_track_y - .3\n",
    "                        va = 'top'\n",
    "                    \n",
    "                    # Draw label\n",
    "                    ax.text(label_x, label_y, hla_gene[4:],\n",
    "                            ha='center', va=va, fontsize=8, \n",
    "                            color='black')\n",
    "                    \n",
    "            else:\n",
    "                rect = Rectangle((gene['start'], gene_track_y), gene_width, 0.6,\n",
    "                                 facecolor='lightgray', alpha=0.3, zorder=1)\n",
    "                ax.add_patch(rect)\n",
    "\n",
    "        # Styling\n",
    "        ax.set_ylabel(f'-log₁₀(P)\\n{label}', fontsize=11)\n",
    "        ax.grid(True, alpha=0.3)\n",
    "        ax.set_xlim(hla_region_genes['start'].min() - 100000,\n",
    "                    hla_region_genes['end'].max() + 100000)\n",
    "        ax.axhline(y=-np.log10(5e-8), color='gray', linestyle='--', alpha=0.75, zorder=0)\n",
    "        ax.axhline(y=-np.log10(bonferroni), color='purple', linestyle='--', alpha=0.75, zorder=2)\n",
    "        \n",
    "        # Set y-axis with gene track below x-axis\n",
    "        ax.set_ylim(-4.5, 23)\n",
    "        ax.set_yticks(np.arange(0, 20.1, 2.5))\n",
    "        \n",
    "        # Despine and remove ticks from all but bottom plot\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        ax.spines['right'].set_visible(False)\n",
    "        ax.spines['left'].set_visible(False)\n",
    "        if idx < len(ancestries) - 1:  # Not the last plot\n",
    "            ax.spines['bottom'].set_visible(False)\n",
    "            ax.tick_params(bottom=False, labelbottom=False)  # Remove x-axis ticks and labels\n",
    "    \n",
    "    # axis formatting (only on last plot)\n",
    "    axes[-1].set_xlabel('Chromosome 6 Position (Mb)', fontsize=12)\n",
    "    axes[-1].xaxis.set_major_formatter(plt.FuncFormatter(lambda x, p: f'{x/1e6:.1f}'))\n",
    "\n",
    "    # Separate legends for each HLA class\n",
    "    class_i_legend = plt.Line2D([0], [0], marker='o', color='w',\n",
    "                               markerfacecolor=class_i_colors[1], markersize=8, alpha=0.8)\n",
    "    class_ii_legend = plt.Line2D([0], [0], marker='o', color='w', \n",
    "                                markerfacecolor=class_ii_colors[1], markersize=8, alpha=0.8)\n",
    "\n",
    "    # Create both legend objects\n",
    "    legend1 = axes[0].legend([class_i_legend], ['HLA Class I (A,B,C)'],\n",
    "                            loc='upper left', frameon=True, fancybox=True)\n",
    "    legend2 = axes[0].legend([class_ii_legend], ['HLA Class II (DR,DQ,DP)'],\n",
    "                            loc='upper right', frameon=True, fancybox=True)\n",
    "\n",
    "    # Add the first legend back as an artist\n",
    "    axes[0].add_artist(legend1)\n",
    "\n",
    "    plt.tight_layout()\n",
    "    return fig, axes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create the plot\n",
    "fig, axes = create_hla_manhattan_plot(data_dict, hla_region_genes, locus_positions, bonf_thresholds)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Forest Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from typing import Tuple\n",
    "\n",
    "def create_hla_forest_plot(\n",
    "    data_dict: dict,\n",
    "    locus_positions: dict,\n",
    "    bonf_thresholds: dict,\n",
    "    max_markers: int = 40,\n",
    "    show_marker_ids: bool = True,\n",
    "    x_min: float = 0.1,\n",
    "    x_max: float = 10,\n",
    ") -> None:\n",
    "    \"\"\"\n",
    "    Create forest plot comparing HLA association results between ancestries.\n",
    "    \"\"\"\n",
    "    \n",
    "    combined_data = []\n",
    "    \n",
    "    for ancestry, (results_df, _) in data_dict.items():\n",
    "        if ancestry == 'metal':\n",
    "            # Handle METAL meta-analysis format\n",
    "            df_with_or = (\n",
    "                results_df\n",
    "                .select([\n",
    "                    \"Marker\",\n",
    "                    pl.col(\"Effect\").alias(\"Beta\"),\n",
    "                    pl.col(\"StdErr\").alias(\"SE\"),\n",
    "                    \"P\"\n",
    "                ])\n",
    "                .with_columns([\n",
    "                    (pl.col(\"P\") < bonf_thresholds[ancestry]).alias(\"bonf_significant\")\n",
    "                ])\n",
    "                .with_columns([\n",
    "                    pl.col(\"Beta\").exp().alias(\"odds_ratio\"),\n",
    "                    (pl.col(\"Beta\") - 1.96 * pl.col(\"SE\")).exp().alias(\"odds_ratio_ci_low\"),\n",
    "                    (pl.col(\"Beta\") + 1.96 * pl.col(\"SE\")).exp().alias(\"odds_ratio_ci_high\"),\n",
    "                    pl.col(\"Marker\").str.split(\".\").list.first().alias(\"HLA_locus\"),\n",
    "                    pl.lit(ancestry.upper()).alias(\"ancestry\")\n",
    "                ])\n",
    "            )\n",
    "        else:\n",
    "            # Handle ancestry-specific format\n",
    "            df_with_or = (\n",
    "                results_df\n",
    "                .select([\"Marker\", \"Beta\", \"SE\", \"P\", \"bonf_significant\"])\n",
    "                .with_columns([\n",
    "                    pl.col(\"Beta\").exp().alias(\"odds_ratio\"),\n",
    "                    (pl.col(\"Beta\") - 1.96 * pl.col(\"SE\")).exp().alias(\"odds_ratio_ci_low\"),\n",
    "                    (pl.col(\"Beta\") + 1.96 * pl.col(\"SE\")).exp().alias(\"odds_ratio_ci_high\"),\n",
    "                    pl.col(\"Marker\").str.split(\".\").list.first().alias(\"HLA_locus\"),\n",
    "                    pl.lit(ancestry.upper()).alias(\"ancestry\")\n",
    "                ])\n",
    "            )\n",
    "        combined_data.append(df_with_or)\n",
    "        \n",
    "    # Combine and process data\n",
    "    all_data = pl.concat(combined_data)\n",
    "    df_pd = all_data.to_pandas()\n",
    "    \n",
    "    # Color scheme by ancestry\n",
    "    ancestry_colors = {\n",
    "        'EUR': '#1f77b4', 'AFR': '#d62728', 'AMR': '#ff7f0e', 'METAL': '#666'\n",
    "    }\n",
    "    \n",
    "    df_pd['color'] = df_pd['ancestry'].map(ancestry_colors).fillna('gray')\n",
    "    df_pd['locus_position'] = df_pd['HLA_locus'].map(locus_positions)\n",
    "    \n",
    "    # Select significant markers and sort\n",
    "    significant_markers = (df_pd.groupby('Marker')['bonf_significant']\n",
    "                          .any().loc[lambda x: x].index.tolist())\n",
    "    df_plot = df_pd[df_pd['Marker'].isin(significant_markers)].copy()\n",
    "    \n",
    "    marker_positions = df_plot.groupby('Marker')['locus_position'].first()\n",
    "    marker_order = (marker_positions.reset_index()\n",
    "                    .sort_values(['locus_position', 'Marker'])['Marker'].tolist())\n",
    "    \n",
    "    # Create figure with dynamic sizing\n",
    "    n_markers = len(marker_order)\n",
    "    figsize = (10, max(6, min(20, n_markers * 0.25 + 3)))\n",
    "    fig, ax = plt.subplots(figsize=figsize)\n",
    "    \n",
    "    # Add backgrounds and group separators\n",
    "    for i in range(len(marker_order)):\n",
    "        if i % 2 == 1:\n",
    "            ax.axhspan(i - 0.5, i + 0.5, color='lightgray', alpha=0.15, zorder=0)\n",
    "\n",
    "    # Add group separators and labels\n",
    "    locus_y_positions = {}\n",
    "    prev_locus = None\n",
    "    \n",
    "    for i, marker in enumerate(marker_order):\n",
    "        locus = marker.split('.')[0]\n",
    "        y_pos = len(marker_order) - i - 1\n",
    "        \n",
    "        # Track locus positions for labels\n",
    "        if locus not in locus_y_positions:\n",
    "            locus_y_positions[locus] = []\n",
    "        locus_y_positions[locus].append(y_pos)\n",
    "        \n",
    "        # Add separator line between groups (extend into label area)\n",
    "        if prev_locus is not None and locus != prev_locus:\n",
    "            y_line = len(marker_order) - i - 0.5\n",
    "            ax.plot([-0.3, 1], [y_line, y_line], color='gray', linestyle='-',\n",
    "                   alpha=0.6, linewidth=0.5, zorder=2, \n",
    "                   transform=ax.get_yaxis_transform(), clip_on=False)\n",
    "        prev_locus = locus\n",
    "        \n",
    "    # Add locus labels\n",
    "    for locus, positions in locus_y_positions.items():\n",
    "        center_y = (max(positions) + min(positions)) / 2\n",
    "        ax.text(-0.3, center_y, f'HLA-{locus}', \n",
    "               transform=ax.get_yaxis_transform(),\n",
    "               ha='right', va='center', fontsize=9, \n",
    "               color='black')\n",
    "\n",
    "    # Create y-positions for plotting\n",
    "    ancestries = df_plot['ancestry'].unique()\n",
    "    offsets = {4: [0.25, 0.08, -0.08, -0.25], 3: [-0.2, 0, 0.2], \n",
    "               2: [-0.15, 0.15]}.get(len(ancestries), [0])\n",
    "    \n",
    "    y_positions = {}\n",
    "    for i, marker in enumerate(marker_order):\n",
    "        base_y = len(marker_order) - i - 1\n",
    "        for j, ancestry in enumerate(sorted(ancestries)):\n",
    "            y_positions[(marker, ancestry)] = base_y + offsets[j % len(offsets)]\n",
    "    \n",
    "    # Plot data points\n",
    "    ancestry_styles = {\n",
    "        'EUR': {'marker': 'o'}, 'AFR': {'marker': 's'}, \n",
    "        'AMR': {'marker': '^'}, 'METAL': {'marker': 'D'}\n",
    "    }\n",
    "    \n",
    "    for _, row in df_plot.iterrows():\n",
    "        y_pos = y_positions.get((row['Marker'], row['ancestry']), 0)\n",
    "        alpha = 1.0 if row['bonf_significant'] else 0.2\n",
    "        style = ancestry_styles.get(row['ancestry'], {'marker': 'o'})\n",
    "        \n",
    "        # Plot CI and point\n",
    "        ax.plot([row['odds_ratio_ci_low'], row['odds_ratio_ci_high']], \n",
    "               [y_pos, y_pos], color=row['color'], linewidth=2, alpha=alpha)\n",
    "        ax.scatter(row['odds_ratio'], y_pos, color=row['color'], \n",
    "                  marker=style['marker'], s=20, alpha=alpha, zorder=5, \n",
    "                  edgecolors='white', linewidth=0.5)\n",
    "    \n",
    "    # Create legend\n",
    "    legend_elements = []\n",
    "    for ancestry in sorted(ancestries):\n",
    "        style = ancestry_styles.get(ancestry, {'marker': 'o'})\n",
    "        color = ancestry_colors.get(ancestry, 'gray')\n",
    "        legend_elements.append(plt.Line2D([0], [0], color=color, linewidth=2, \n",
    "                                        marker=style['marker'], markersize=6, \n",
    "                                        alpha=1.0, label=f'{ancestry}'))\n",
    "    \n",
    "    legend_elements.extend([\n",
    "        plt.Line2D([0], [0], color='gray', linewidth=2, alpha=1.0, \n",
    "                   label='Bonferroni Significant'),\n",
    "        plt.Line2D([0], [0], color='gray', linewidth=2, alpha=0.2, \n",
    "                   label='Non-significant')\n",
    "    ])\n",
    "\n",
    "    # Customize plot\n",
    "    ax.set_xscale('log')\n",
    "    ax.axvline(x=1, color='black', linestyle='-', alpha=0.8, linewidth=2, zorder=1)\n",
    "    \n",
    "    # Add reference lines\n",
    "    for ref_or in [0.5, 1.5, 2, 3, 5]:\n",
    "        if x_min <= ref_or <= x_max:\n",
    "            ax.axvline(x=ref_or, color='lightgray', linestyle=':', \n",
    "                      alpha=0.5, linewidth=1)\n",
    "    \n",
    "    # Format axes\n",
    "    ax.set_ylim(-0.5, len(marker_order) - 0.5)\n",
    "    ax.set_yticks(range(len(marker_order)))\n",
    "    ax.set_yticklabels(list(reversed(marker_order)), fontsize=9)\n",
    "    \n",
    "    ax.set_xlim(x_min, x_max)\n",
    "    x_ticks = [0.1, 0.2, 0.5, 1, 2, 5, 10, 20, 50]\n",
    "    x_ticks_filtered = [tick for tick in x_ticks if x_min <= tick <= x_max]\n",
    "    ax.set_xticks(x_ticks_filtered)\n",
    "    ax.set_xticklabels([str(tick) for tick in x_ticks_filtered])\n",
    "    \n",
    "    # Labels and styling\n",
    "    ax.set_xlabel('Odds Ratio', fontsize=10)\n",
    "    ax.set_ylabel('', fontsize=12)\n",
    "    ax.set_title(f'HLA Association Forest Plot\\n({len(marker_order)} Alleles Significant in at Least One Group)', \n",
    "                fontsize=10)\n",
    "    \n",
    "    ax.legend(handles=legend_elements, bbox_to_anchor=(1.05, 1), \n",
    "             loc='upper left', fontsize=8)\n",
    "    ax.grid(True, alpha=0.3, axis='x', which='both')\n",
    "    ax.set_axisbelow(True)\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.spines['right'].set_visible(False)\n",
    "    \n",
    "    plt.tight_layout()\n",
    "    return fig, ax"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = create_hla_forest_plot(data_dict, locus_positions, bonf_thresholds, x_min=0.1, x_max=5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Vulvovaginal Disease"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## HLA and GWAS Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "afr_vv_hla_df = pl.read_csv(f'{my_bucket}/HLA/HPV_AFR_VV_HLA_results_freq.csv')\n",
    "eur_vv_hla_df = pl.read_csv(f'{my_bucket}/HLA/HPV_EUR_VV_HLA_results_freq.csv')\n",
    "amr_vv_hla_df = pl.read_csv(f'{my_bucket}/HLA/HPV_AMR_VV_HLA_results_freq.csv')\n",
    "metal_vv_hla_df = pl.read_csv(f'{my_bucket}/HLA/meta_VV_1.txt', separator='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "eur_vv_sumstats = gl.Sumstats(f'{my_bucket}/saige_gwas/v2/eur/condition__hpv/gwas_results.tsv.gz',\n",
    "                           fmt=\"saige\",\n",
    "                           snpid='vid',\n",
    "                           rsid=None,\n",
    "                           chrom='chromosome',\n",
    "                           pos='base_pair_location',\n",
    "                           ea='effect_allele',\n",
    "                           nea='other_allele',\n",
    "                           eaf='effect_allele_frequency',\n",
    "                           se='standard_error',\n",
    "                           p='p_value',\n",
    "                           beta='beta',\n",
    "                           build='38'\n",
    ")\n",
    "eur_vv_sumstats.basic_check()\n",
    "eur_hla_vv_sumstats = pl.from_pandas(eur_vv_sumstats.data).filter(\n",
    "    (pl.col('CHR')==6) &\n",
    "    ((pl.col('POS')>hla_start) | (pl.col('CHR')<hla_end))\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "afr_vv_sumstats = gl.Sumstats(f'{my_bucket}/saige_gwas/v2/afr/condition__hpv/gwas_results.tsv.gz',\n",
    "                           fmt=\"saige\",\n",
    "                           snpid='vid',\n",
    "                           rsid=None,\n",
    "                           chrom='chromosome',\n",
    "                           pos='base_pair_location',\n",
    "                           ea='effect_allele',\n",
    "                           nea='other_allele',\n",
    "                           eaf='effect_allele_frequency',\n",
    "                           se='standard_error',\n",
    "                           p='p_value',\n",
    "                           beta='beta',\n",
    "                           build='38'\n",
    ")\n",
    "afr_vv_sumstats.basic_check()\n",
    "afr_hla_vv_sumstats = pl.from_pandas(afr_vv_sumstats.data).filter(\n",
    "    (pl.col('CHR')==6) &\n",
    "    ((pl.col('POS')>hla_start) | (pl.col('CHR')<hla_end))\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "amr_vv_sumstats = gl.Sumstats(f'{my_bucket}/saige_gwas/v2/amr/condition__hpv/gwas_results.tsv.gz',\n",
    "                           fmt=\"saige\",\n",
    "                           snpid='vid',\n",
    "                           rsid=None,\n",
    "                           chrom='chromosome',\n",
    "                           pos='base_pair_location',\n",
    "                           ea='effect_allele',\n",
    "                           nea='other_allele',\n",
    "                           eaf='effect_allele_frequency',\n",
    "                           se='standard_error',\n",
    "                           p='p_value',\n",
    "                           beta='beta',\n",
    "                           build='38'\n",
    ")\n",
    "amr_vv_sumstats.basic_check()\n",
    "amr_hla_vv_sumstats = pl.from_pandas(amr_vv_sumstats.data).filter(\n",
    "    (pl.col('CHR')==6) &\n",
    "    ((pl.col('POS')>hla_start) | (pl.col('CHR')<hla_end))\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_chr_pos_from_snpid(df: pd.DataFrame, snp_col: str = \"SNPID\") -> pd.DataFrame:\n",
    "    \"\"\"Extract CHR and POS from SNPID column (format CHR-POS-...).\"\"\"\n",
    "    parts = df[snp_col].str.split(\"-\", n=2, expand=True)\n",
    "    df = df.copy()\n",
    "    df[\"CHR\"] = pd.to_numeric(parts[0], errors=\"coerce\")\n",
    "    df[\"POS\"] = pd.to_numeric(parts[1], errors=\"coerce\")\n",
    "    return df\n",
    "\n",
    "metal_vv_sumstats = gl.Sumstats(f'{my_bucket}/saige_gwas/v2/metal/condition__hpv/condition__hpv_afr_amr_eur1.tsv',\n",
    "                           fmt=\"metal\",\n",
    "                           snpid='MarkerName',\n",
    "                           rsid=None,\n",
    "                           ea='Allele1',\n",
    "                           nea='Allele2',\n",
    "                           eaf='Freq1',\n",
    "                           se='StdErr',\n",
    "                           p='P-value',\n",
    "                           beta='Effect',\n",
    "                           direction='Direction',\n",
    "                           build='38'\n",
    ")\n",
    "# Get CHR POS from SNPID\n",
    "metal_vv_sumstats.data = add_chr_pos_from_snpid(metal_vv_sumstats.data, snp_col='SNPID')\n",
    "metal_vv_sumstats.basic_check()\n",
    "metal_hla_vv_sumstats = pl.from_pandas(metal_vv_sumstats.data).filter(\n",
    "    (pl.col('CHR')==6) &\n",
    "    ((pl.col('POS')>hla_start) | (pl.col('CHR')<hla_end))\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metal_vv_hla_df = metal_vv_hla_df.rename({'MarkerName':'Marker', 'P-value': 'P'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hla_genes_positions = hla_region_genes.filter(pl.col('name').str.contains('HLA')).sort('name')\n",
    "    \n",
    "hla_genes_positions = hla_genes_positions.with_columns(\n",
    "    pl.col(\"name\").str.split(\"-\").list.last().alias(\"hla_name\"),\n",
    ")\n",
    "\n",
    "locus_positions = dict(zip(hla_genes_positions['hla_name'], hla_genes_positions['mid_point']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove DMA and DRA from analysis because AC == 1 for both\n",
    "eur_vv_hla_df = eur_vv_hla_df.filter(~pl.col('Marker').str.contains('DMA|DRA'))\n",
    "afr_vv_hla_df = afr_vv_hla_df.filter(~pl.col('Marker').str.contains('DMA|DRA'))\n",
    "amr_vv_hla_df = amr_vv_hla_df.filter(~pl.col('Marker').str.contains('DMA|DRA'))\n",
    "metal_vv_hla_df = metal_vv_hla_df.filter(~pl.col('Marker').str.contains('DMA|DRA'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Prepare your data dictionary\n",
    "data_dict = {\n",
    "    'eur': (eur_vv_hla_df, eur_hla_vv_sumstats),\n",
    "    'afr': (afr_vv_hla_df, afr_hla_vv_sumstats), \n",
    "    'amr': (amr_vv_hla_df, amr_hla_vv_sumstats),\n",
    "    'metal': (metal_vv_hla_df, metal_hla_vv_sumstats)\n",
    "}\n",
    "\n",
    "# Calculate Bonferroni thresholds\n",
    "bonf_thresholds = {\n",
    "    'eur': 0.05 / len(eur_vv_hla_df),\n",
    "    'afr': 0.05 / len(afr_vv_hla_df),\n",
    "    'amr': 0.05 / len(amr_vv_hla_df),\n",
    "    'metal': 0.05 / len(metal_vv_hla_df)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### GWAS Overlay"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create the plot\n",
    "fig, axes = create_hla_manhattan_plot(data_dict, hla_region_genes, locus_positions, bonf_thresholds)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Forest Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = create_hla_forest_plot(data_dict, locus_positions, bonf_thresholds, x_min=0.1, x_max=5)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "210.594px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
